R/load.R

```diff
@@ -388,11 +383,24 @@ is_loading <- function(pkg = NULL) {
   }
 }
 
-# Ensure that calls to `::` resolve to the original unregistered
-# namespace
+# Ensure that calls to `::` resolve to the original detached namespace. It is
+# uncommon to refer to functions in your own package with `::` but it sometimes
+# is necessary, e.g. in standalone files.
+#
+# To enable this, assign `::` in your namespace, e.g.
+#
+# ```
+# on_load(`::` <- base::`::`)
+# ```
 patch_colon <- function(package) {
   ns <- asNamespace(package)
-  rlang::env_unlock(ns)
+
+  if (!env_has(ns, "::")) {
+    return()
```

Comment:
It feels like it would be useful to log this in some way, but if you think it only affects rlang, it's probably fine as is.

Reply:
It's only used in rlang AFAIK and even there I think we ran into other issues that prevented us from using `::` in some standalone files. So I prefer to keep this feature low-key, and a warning would affect all pkgload users. But you're right that if this feature does end up being needed by someone, they will have a very hard time to discover this fix. We can revisit if someone reports an issue about this.